tracer/src/Datadog.Trace/Telemetry/Collectors/MetricsTelemetryCollector.cs

```diff
@@ -0,0 +1,478 @@
+                {
+                    data.Add(
+                        new DistributionMetricData(
+                            metricName,
+                            points: queue.ToArray(),
+                            common: metric.IsCommon()));
+                }
+            }
+        }
+
+        foreach (var kvp in distributionsWithTags)
+        {
+            var helper = kvp.Key;
+            var metric = (Distribution)helper.Metric;
+            if (metric.GetName() is { } metricName)
+            {
+                data.Add(
+                    new DistributionMetricData(
+                        metricName,
+                        points: kvp.Value.ToArray(),
+                        common: metric.IsCommon()) { Tags = helper.GetTags(), });
+            }
+        }
+
+        return data;
+    }
+
+    private void RecordCount(in MetricKey key, int increment)
+    {
+#if NETCOREAPP
+        // we can avoid the closure in .NET Core
+        _buffer.CountsWithTags.AddOrUpdate(
+            key: key,
+            addValueFactory: static (_, arg) => arg,
+            updateValueFactory: static (_, old, arg) => (old + arg) < old ? int.MaxValue : (old + arg),
+            factoryArgument: increment);
+#else
+        _buffer.CountsWithTags.AddOrUpdate(
+            key,
+            increment,
+            (_, old) => (old + increment) < old ? int.MaxValue : (old + increment)); // look out for overflow
+#endif
+    }
+
+    private void RecordGauge(in MetricKey key, int value)
+    {
+#if NETCOREAPP
+        // we can avoid the closure in .NET Core
+        _buffer.GaugesWithTags.AddOrUpdate(
+            key: key,
+            addValueFactory: static (_, arg) => arg,
+            updateValueFactory: static (_, _, arg) => arg,
+            factoryArgument: value);
+#else
+        _buffer.GaugesWithTags.AddOrUpdate(key, value, (_, old) => value);
+#endif
+    }
+
+    private void RecordDistribution(in MetricKey key, double value)
+    {
+#if NETCOREAPP
+        // we can avoid the closure in .NET Core
+        _buffer.DistributionsWithTags.AddOrUpdate(
+            key: key,
+            addValueFactory: static (_, arg) =>
+            {
+                var q = MetricBuffer.CreateDistributionQueue();
+                q.TryEnqueue(arg);
+                return q;
+            },
+            updateValueFactory: static (_, queue, arg) =>
+            {
+                queue.TryEnqueue(arg);
+                return queue;
+            },
+            factoryArgument: value);
+#else
+        _buffer.DistributionsWithTags.AddOrUpdate(
+            key,
+            _ =>
+            {
+                var queue = MetricBuffer.CreateDistributionQueue();
+                queue.TryEnqueue(value);
+                return queue;
+            },
+            (_, queue) =>
+            {
+                queue.TryEnqueue(value);
+                return queue;
+            });
+#endif
+    }
```

Comment:
Since you're not actually updating the value, I think it's better to do:

```suggestion
        var queue = _buffer.DistributionsWithTags.GetOrAdd(key, _ => MetricBuffer.CreateDistributionQueue());
        queue.TryEnqueue(value);
    }
```
Unless it's an issue that another thread sees an empty queue.

Reply:
> Unless it's an issue that another thread sees an empty queue

Well, that is _theoretically_ an issue, as we could technically lose this value if it's written while the polling is happening. In practice though, there's lots of edge cases like that in here, so I think the tradeoff is good enough